rpmduplicates.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RPMDUPLICATES 1"
.TH RPMDUPLICATES 1 "2012-08-01" "perl v5.14.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
rpmduplicates \- find program with several version installed
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
sometimes, the upgrade of a system will install new packages, and \*(L"forget\*(R" to remove
the older version. Rpmorphan do not work with this kind of unused packages. Rpmduplicates 
is the answer. I recommend to run it after each distribution upgrade 
(for example from fedora core 7 to 8).
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
rpmduplicates.pl  [options]
.PP
options:
.PP
.Vb 3
\&   \-help                brief help message
\&   \-man                 full documentation
\&   \-V, \-\-version        print version
\&
\&   \-verbose             verbose
.Ve
.SH "REQUIRED ARGUMENTS"
.IX Header "REQUIRED ARGUMENTS"
nothing
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-help\fR" 8
.IX Item "-help"
Print a brief help message and exits.
.IP "\fB\-man\fR" 8
.IX Item "-man"
Print the full manual page and exits.
.IP "\fB\-version\fR" 8
.IX Item "-version"
Print the program release and exit.
.IP "\fB\-verbose\fR" 8
.IX Item "-verbose"
The program works and print debugging messages.
.SH "USAGE"
.IX Header "USAGE"
rpmduplicates.pl
.SH "FILES"
.IX Header "FILES"
nothing
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
it will show the duplicates with their version and build date
.PP
and suggest to remove the older one
.PP
example :
.PP
duplicate kernel-devel : 2.6.24.4\-64.fc8 (Sat Mar 29 15:36:41 2008) / 2.6.24.3\-50.fc8 (Thu Mar 20 20:43:47 2008)
.PP
suggest : rpm \-e kernel\-devel\-2.6.24.3\-50.fc8
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
the number of duplicates packages found
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
nothing
.SH "DEPENDENCIES"
.IX Header "DEPENDENCIES"
rpmorphan library
.SH "INCOMPATIBILITIES"
.IX Header "INCOMPATIBILITIES"
not known
.SH "BUGS AND LIMITATIONS"
.IX Header "BUGS AND LIMITATIONS"
not known
.SH "NOTES"
.IX Header "NOTES"
this program can be used as \*(L"normal\*(R" user
.PP
the gpg-pubkey packages are excluded from the result
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\fIrpm\fR\|(1) for rpm call
.PP
\fIrpmorphan\fR\|(1)
.PP
\fIrpmusage\fR\|(1)
.PP
\fIrpmdep\fR\|(1)
.PP
\fIrpmextra\fR\|(1)
.SH "LICENSE AND COPYRIGHT"
.IX Header "LICENSE AND COPYRIGHT"
Copyright (C) 2008 by Eric Gerbier
This program is free software; you can redistribute it and/or modify
it under the terms of the \s-1GNU\s0 General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH "AUTHOR"
.IX Header "AUTHOR"
Eric Gerbier
.PP
you can report any bug or suggest to gerbier@users.sourceforge.net


NEWS
			1.11 version
			-------------
new features :
* add suggests option
* add /etc/rpmorphanrc
* add env RPMORPHAN_METHOD for debugging

bugfix :
* fix bug with space in filenames
-------------------------------------------------------------------------------
			1.10 version
			-------------
new features :
* display package summary in a popup on S key

bugfix :
* bugfix eval for Curses (mandriva)
* fix bad version display

improvment :
* reduce size/duplicated code in rpmorphan/rpmusage
* reduce menu length in curses gui (< 72)

others :
* clean all scripts
-------------------------------------------------------------------------------
			1.9 version
			-------------
new features :
- add rpmextra tool
- add grpmorphan as link to rpmorphan -gui
others :
- fix perlcritic warnings
-------------------------------------------------------------------------------
			1.8 version
			-------------
bugfix : 
- fix bug on rpmdep ( missing installtime in urpm mode)

new features :
- add help translation (gui)
-------------------------------------------------------------------------------
			1.7 version
			-------------
bugfix : 
- Curses is no more an rpm dependency

new features : 
- add reload button
-------------------------------------------------------------------------------
			1.6 version
			-------------
other :
- add a confirm dialog box on remove action
- curses window size is now variable to use full available size
- info window is now scrollable (curses gui)
-------------------------------------------------------------------------------
			1.5 version
			-------------
new features : 
- use URPM perl package for faster database access on mandriva
- can use RPM2 perl package for faster database access (fedora)
bugfix : 
- fix a bug on curses interface
- remove all matching packages
others :
- merge spec file with fedora one
- add info for fedora/mandriva users
-------------------------------------------------------------------------------
			1.4 version
			-------------
other :
when removing a package, anf if rpm send some output, the code now check the rpm 
database to know if the remove works or not
-------------------------------------------------------------------------------
			1.3 version
			-------------
incompatibility : none
bugfix : 
- fix a bug if exclude are set in config file (thanks Szymon Siwek)
new features : 
- display number of deleted file
other :
-------------------------------------------------------------------------------
			1.2 version
			-------------
incompatibility : none
bugfix : none
new features : 
- write log to /var/log/rpmorphan.log
- add rpmdep.pl tool
- add rpmduplicate.pl tool
other :
-------------------------------------------------------------------------------
			1.1 version
			-------------
incompatibility : none
bugfix : none
new features : 
- (rpmorphan) add a default target : guess-lib
- (rpmorphan) add keep-file option, which allow to cutomize the keep file
- (rpmusage) add a default target : all
other :
- the different tools now share a common library
- rpmlint was used to improve the rpm package

-------------------------------------------------------------------------------
			1.0 version
			-------------
add dry-run (simulation) mode
(gui) add 3 counters : packages, orphans, selected
(gui) recursive analysis
improve virtuals : now check if how many package provide this one
-------------------------------------------------------------------------------
			0.9 version
			-------------
add another graphical interface (curses), available by the --curses command line
argument
add configuration file
-------------------------------------------------------------------------------
			0.8 version
			-------------
add simple graphical user interface (optionnal), which require
the perl-tk package :use --gui command line argument

the code has been split in smaller subroutines.
-------------------------------------------------------------------------------
			0.4 version
			-------------
add optionnal cache, which can be shared by all rpmorphan tools
to spare rpm query time :
use --use-cache and --clear-cache command line arguments
-------------------------------------------------------------------------------
			0.3 version
			-------------
new feature : add rpmusage tool
-------------------------------------------------------------------------------

			0.2 version
			-------------
new feature : add a permanent list of exclude file (default is empty) and
4 new options for rpmorphan : --zero-keep, --list-keep, --add-keep, --del-keep

-------------------------------------------------------------------------------
			0.1 version
			-------------

first public release


rpmduplicates.pl
#!/usr/bin/perl
###############################################################################
#
#    Copyright (C) 2006 by Eric Gerbier
#    Bug reports to: gerbier@users.sourceforge.net
#    $Id: rpmorphan-1.11 | rpmduplicates.pl | Mon Jul 4 12:47:55 2011 +0000 | gerbier  $
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
###############################################################################
# search for packages with several versions installed
###############################################################################
use strict;
use warnings;

use English '-no_match_vars';

use Getopt::Long;    # arg analysis
use Pod::Usage;      # man page
use Data::Dumper;    # debug

# library
use File::Basename;
my $dirname = dirname($PROGRAM_NAME);

# not very usefull, but warning, debug ...
require $dirname . '/rpmorphan-lib.pl'
  or die "problem to load rpmorphan-lib: $ERRNO\n";

my $VERSION = '0.2';

my $opt_help;
my $opt_man;
my $opt_version;
my $opt_verbose;

my %opt = (
	'help'    => \$opt_help,
	'man'     => \$opt_man,
	'verbose' => \$opt_verbose,
	'version' => \$opt_version,
);

Getopt::Long::Configure('no_ignore_case');
GetOptions( \%opt, 'help|?', 'man', 'verbose', 'version|V' )
  or pod2usage(2);

init_debug($opt_verbose);

if ($opt_help) {
	pod2usage(1);
}
elsif ($opt_man) {
	pod2usage( -verbose => 2 );
}
elsif ($opt_version) {
	print_version($VERSION);
	exit;
}

#  we cannot use common cache as duplicate package are not clean taken
#  by the common code
## no critic (RequireInterpolationOfMetachars)
my $cmd =
  'rpm -qa --queryformat "%{NAME} %{VERSION}-%{RELEASE} %{BUILDTIME}\n" ';
## use critic

my $fh;
## no critic (ProhibitTwoArgOpen,RequireBriefOpen)
if ( !open $fh, "$cmd |" ) {
	die "can not open $cmd : $ERRNO\n";
}
## use critic

my %h_ver;          # versions
my %h_buildtime;    # buildtime

my $nb    = 0;
my $SPACE = q{ };

while ( my $line = <$fh> ) {
	chomp $line;
	my ( $soft, $version, $build ) = split $SPACE, $line;

	next if ( $soft eq 'gpg-pubkey' );

	if ( exists $h_ver{$soft} ) {

		# we already have found a package with same name
		my $old_ver  = $h_ver{$soft};
		my $old_date = $h_buildtime{$soft};

		my $old;
		if ( $old_date < $build ) {

			# current package was build after the stored one
			# keep the latest
			$h_buildtime{$soft} = $build;

			# delete the older
			$old = $old_ver;
		}
		else {

			# current package was build before the stored one
			# delete the older
			$old = $version;
		}
		$old_date = localtime $old_date;
		$build    = localtime $build;
		print "duplicate $soft : $old_ver ($old_date) / $version ($build)\n";
		print "suggest : rpm -e $soft-$old\n";
		$nb++;
	}
	else {
		$h_ver{$soft}       = $version;
		$h_buildtime{$soft} = $build;
	}
}
close $fh or warning("can not close $cmd : $ERRNO");

if ( $nb == 0 ) {
	print "no duplicate found !\n";
}

exit $nb;
__END__

=head1 NAME

rpmduplicates - find program with several version installed

=head1 DESCRIPTION

sometimes, the upgrade of a system will install new packages, and "forget" to remove
the older version. Rpmorphan do not work with this kind of unused packages. Rpmduplicates 
is the answer. I recommend to run it after each distribution upgrade 
(for example from fedora core 7 to 8).

=head1 SYNOPSIS

rpmduplicates.pl  [options]

options:

   -help                brief help message
   -man                 full documentation
   -V, --version        print version

   -verbose             verbose

=head1 REQUIRED ARGUMENTS

nothing

=head1 OPTIONS

=over 8

=item B<-help>

Print a brief help message and exits.

=item B<-man>

Print the full manual page and exits.

=item B<-version>

Print the program release and exit.

=item B<-verbose>

The program works and print debugging messages.

=back

=head1 USAGE

rpmduplicates.pl 

=head1 FILES

nothing

=head1 DIAGNOSTICS

it will show the duplicates with their version and build date

and suggest to remove the older one

example :

duplicate kernel-devel : 2.6.24.4-64.fc8 (Sat Mar 29 15:36:41 2008) / 2.6.24.3-50.fc8 (Thu Mar 20 20:43:47 2008)

suggest : rpm -e kernel-devel-2.6.24.3-50.fc8

=head1 EXIT STATUS

the number of duplicates packages found

=head1 CONFIGURATION

nothing

=head1 DEPENDENCIES

rpmorphan library

=head1 INCOMPATIBILITIES

not known

=head1 BUGS AND LIMITATIONS

not known

=head1 NOTES

this program can be used as "normal" user

the gpg-pubkey packages are excluded from the result

=head1 SEE ALSO

=for man
\fIrpm\fR\|(1) for rpm call
.PP
\fIrpmorphan\fR\|(1)
.PP
\fIrpmusage\fR\|(1)
.PP
\fIrpmdep\fR\|(1)
.PP
\fIrpmextra\fR\|(1)

=for html
<a href="rpmorphan.1.html">rpmorphan(1)</a><br />
<a href="rpmusage.1.html">rpmusage(1)</a><br />
<a href="rpmdep.1.html">rpmdep(1)</a><br />
<a href="rpmextra.1.html">rpmextra(1)</a><br />

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2008 by Eric Gerbier
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

=head1 AUTHOR

Eric Gerbier

you can report any bug or suggest to gerbier@users.sourceforge.net

=cut

